### PR TITLE
[AutoDiff] Eliminate SIL linking from differentiation.

### PR DESCRIPTION
**Pull request by dan-zheng**
The differentiation pass no longer depends on SerializedSILLoader to link
JVP/VJP functions or witness tables.

Background: the differentiation pass guarantees that JVP/VJPs exist for
`@differentiable` functions defined in a module. If the original function
has public linkage, then JVP/VJPs also have public linkage.

If JVP/VJPs cannot be found in the current module for a given original
function, create an empty declaration to them with public external linkage.
No deserialization is necessary.

A similar technique has been used for witness methods: we emit
`witness_method` instructions rather than loading witness tables.
"Defensive" explicit loading of witness tables is removed.

---

This enables differentiation of functions from other modules, necessary for
library development.

Joint work with @rxwei.

Old thread about linking witness tables:
https://forums.swift.org/t/how-to-link-one-witness-table-from-stdlib/12674

### CODE REVIEW

**Comment by rxwei** [comment]
> This enables differentiation of functions from other modules, necessary for
library development.

It'd be more accurate to say "This enables differentiation of functions from other modules **without requiring functions to be `@inlinable`**".

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -3639,25 +3630,89 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
...
+      auto jvpType = originalTy->getAutoDiffAssociatedFunctionType(
+          getIndices().parameters, getIndices().source, 1,
+          AutoDiffAssociatedFunctionKind::JVP, module,
+          LookUpConformanceInModule(module.getSwiftModule()), jvpGenericSig);
+      SILOptFunctionBuilder fb(context.getTransform());
+      jvp = fb.createFunction(SILLinkage::PublicExternal, jvpName, jvpType,
+                              jvpGenericEnv, originalLoc, original->isBare(),
+                              IsNotTransparent, original->isSerialized());
+      // NOTE: Setting debug scope is necessary to prevent crash in TFPartition.
+      jvp->setDebugScope(new (module) SILDebugScope(originalLoc, jvp));
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: To elaborate: [`TensorFunctionClassifier::shouldBePartitioned(SILFunction *fn, bool forceTFFunctions)`](https://github.com/apple/swift/blob/b1a9dae019604daa9525a6e6a4718a643178cdee/lib/SILOptimizer/Mandatory/TFUtilities.cpp#L401) calls `fn->getDeclContext()` without first checking whether `fn` has a valid debug scope, causing a crash.

However, `SILFunction`s should have debug scopes, so it's reasonable to assume they do have one.

**Review by rxwei — APPROVED** [review]
Awesome! When you get a chance, please also send a patch to the const evaluator so that we can revert our earlier change that made `SILModule::getSILLoder()` public.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -3639,25 +3630,89 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
...
+  // Add where clause requirements.
+  auto source =
+      GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+  for (auto &req : attr->getRequirements())
+    builder.addRequirement(req, source, original->getModule().getSwiftModule());
+  auto canGenericSig = std::move(builder)
+                           .computeGenericSignature(
+                               SourceLoc(), /*allowConcreteGenericParams=*/true)
+                           ->getCanonicalSignature();
+  return canGenericSig;
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: The return type of this method should either be a CanGenericSignature, or you should not call getCanonicalSignature().

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -3639,25 +3630,89 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
...
   if (attr->hasVJP()) {
-    vjp = lookUpOrLinkFunction(attr->getVJPName(), context.getModule());
-    assert(vjp);
+    // If attribute specifies VJP name, try to look up VJP in current module.
+    // Otherwise, create an external reference.
+    vjp = module.lookUpFunction(attr->getVJPName());
+    if (!vjp) {
+      auto vjpName = attr->getVJPName();
+      auto *vjpGenericSig = getAutoDiffAssociatedFunctionGenericSignature(attr, original);
+      auto *vjpGenericEnv = vjpGenericSig
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: SILFunctions without a body don't need a GenericEnvironment

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks! Done in https://github.com/apple/swift/pull/21973/commits/41716eb58db03ad360f3bafec46e16bf000050d1.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done in https://github.com/apple/swift/pull/21973/commits/41716eb58db03ad360f3bafec46e16bf000050d1.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Pull request merged**